### PR TITLE
Improve task timeout reliability and diagnostics

### PR DESCRIPTION
**Pull request by ldaley**
This has two functional changes:

1. When a work unit times out, we know continue to interrupt it until it stops. Previously we did this only once.
2. When a work unit is slow to stop, we now log its stack trace every 10s if it has changed since last being logged

Here's an example of the output for part 2:

```

> Task :block
Requesting stop of task ':block' as it has exceeded its configured timeout of 500ms.
Timed out task ':block' has not yet stopped.
Current stacktrace of timed out but not yet stopped task ':block':
  at build_67ste90xdw0dy37opt2wkr8tj.checkpoint1(/Users/ld/Projects/gradle/subprojects/core/build/tmp/test files/TaskTimeout.Test/e9z0g/build.gradle:5)
  at java.base@11.0.4/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base@11.0.4/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

Timed out task ':block' has not yet stopped.
Timed out task ':block' has not yet stopped.
Timed out task ':block' has not yet stopped.
Timed out task ':block' has not yet stopped.
Current stacktrace of timed out but not yet stopped task ':block':
  at build_67ste90xdw0dy37opt2wkr8tj.checkpoint2(/Users/ld/Projects/gradle/subprojects/core/build/tmp/test files/TaskTimeout.Test/e9z0g/build.gradle:10)
  at java.base@11.0.4/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base@11.0.4/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

Timed out task ':block' has not yet stopped.
Timed out task ':block' has not yet stopped.
Current stacktrace of timed out but not yet stopped task ':block':
  at build_67ste90xdw0dy37opt2wkr8tj.checkpoint3(/Users/ld/Projects/gradle/subprojects/core/build/tmp/test files/TaskTimeout.Test/e9z0g/build.gradle:15)
  at java.base@11.0.4/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  at java.base@11.0.4/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

Timed out task ':block' has not yet stopped.
Timed out task ':block' has not yet stopped.
Timed out task ':block' has stopped.

> Task :block FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':block'.
> Timeout has been exceeded
```

### CODE REVIEW

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Review by adammurdoch — APPROVED** [review]

**Comment by ldaley** [comment]
@bot-gradle test RFM

**Comment by ldaley** [comment]
@bot-gradle test RFM

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy [hunk 1]
```diff
@@ -288,4 +294,86 @@ class TaskTimeoutIntegrationTest extends AbstractIntegrationSpec {
...
+    }
+
+    def "stack trace of task is printed if it is slow to stop"() {
+        given:
+        executer.withStackTraceChecksDisabled()
+        postTimeoutCheckFrequencyMs = 100
+        slowStopLogStacktraceFrequencyMs = 100
+
+        buildFile << """
+            @groovy.transform.CompileStatic
```
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy:334]: Why the `CompileStatic`? It makes the test harder to read.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java [hunk 2]
```diff
@@ -76,35 +87,73 @@ private static long warnIfNotStoppedFrequency() {
...
+
+                logMessageWriter.print("Current stacktrace of timed out but not yet stopped ");
+                logMessageWriter.print(workUnitDescription.getDisplayName());
+                logMessageWriter.println(":");
+
+                for (StackTraceElement traceElement : currentStackTrace) {
+                    logMessageWriter.println("  at " + traceElement);
                 }
+
+                logMessageWriter.close();
```
Comment (wolfs) [inline subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java:154]: You could use try with resources to close the writer - or not close it at all.

**Review by wolfs — APPROVED** [review]
LGTM! I left a small comment.

File: subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy (hunk 1 above)
Comment (ldaley) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy:334]: It reduces the volatility in the traces, making the test output/results a bit easier to reason about.

Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy:334]: 👍 

Does that maybe mean we should filter the stacktraces we print out for the threads so we remove the groovy dynamic invocation parts?

Comment (ldaley) [reply subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy:334]: I tried that, but it's a bit involved. This module can't see the filterer.

File: subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java (hunk 2 above)
Comment (ldaley) [reply subprojects/execution/src/main/java/org/gradle/internal/execution/timeout/impl/DefaultTimeoutHandler.java]: Went for try-with-resources.

**Review by wolfs — APPROVED** [review]
LGTM!

**Pull request merged**